activejob/lib/active_job/exceptions.rb

```diff
@@ -25,8 +25,9 @@ module ClassMethods
       #   as a computing proc that takes the number of executions so far as an argument, or as a symbol reference of
       #   <tt>:exponentially_longer</tt>, which applies the wait algorithm of <tt>((executions**4) + (Kernel.rand * (executions**4) * jitter)) + 2</tt>
       #   (first wait ~3s, then ~18s, then ~83s, etc)
-      # * <tt>:attempts</tt> - Re-enqueues the job the specified number of times (default: 5 attempts) or a symbol reference of <tt>:unlimited</tt>
-      #   to retry the job until it succeeds
+      # * <tt>:attempts</tt> - Re-enqueues the job either the specified number of times (default: 5 attempts),
```

Comment:
I wonder if it is clear from the name that it is meant to return a boolean? I could see someone returning 0 and running into trouble because it is truthy.

Would it make sense to have an alternate mutually exclusive parameter with a different name, rather than reusing `attempts`? Something like `attempt_again` or `reattempt`?

Reply:
Oh, that is a good point. It does become less clear. Originally I was just going to make the proc return a number to be compared... but then I thought it would be good to add more flexibility by making it return true/false.

So two questions:
- What do you think of still allowing a Proc for `attempts` but making that proc return a number so it still gives flexibility to generate that number from something instead of passing in an Int?
- If that happens do you think I should also add support for :reattempt that is true/false? I don't need it for what I'm doing but I'd be happy to add it for the greater flexibility for others in the future. And if I do should it be part of this PR or a separate one?

Reply:
Hi @rosew great PR! I just opened https://github.com/rails/rails/pull/46261 and was pointed over to this. I'd be in strong support of:

> What do you think of still allowing a Proc for attempts but making that proc return a number so it still gives flexibility to generate that number from something instead of passing in an Int?

That would give good flexibility for most cases imo. 

In https://github.com/rails/rails/pull/46261 I had intended to pass the job instance to the attempts proc, to allow callers to dynamically specify the desired number of attempts by doing something like 
```ruby
class MyJob < ActiveJob::Base
    retry_on CustomRetryException, attempts: ->(job){ job.arguments.last[:retries] || 5 }
    def perform(retries:)
        # job implementation
    end
end
MyJob.perform_later(retries: 10)
MyJob.perform_later(retries: 1)
```
What are your thoughts on doing something like that?

Reply:
The main reason I passed in executions was because it mirrored the behavior of the other Proc accepted for delay. I don't have a strong opinion either way on if this would be a good case to break from the current pattern or not. 

I do see how my method can allow it to be customized at the Class level where yours can allow it to be customized at the instance level.

I'd love thoughts from rails maintainers on how close we should stick to the other proc pattern and/or if there are reasons not to pass the into the proc.